src/app.rs

```diff
@@ -199,12 +204,93 @@ impl App {
         Ok(app)
     }
 
+    async fn restore_or_initialize_tree(
+        database: &Database,
+        root_hash: ruint::Uint<256, 4>,
+        tree_depth: usize,
+        dense_prefix_depth: usize,
+        gc_threshold: usize,
+        initial_leaf_value: Hash,
+        mmap_file_path: String,
+    ) -> AnyhowResult<TreeState> {
+        let tree_state = match Self::get_cached_tree_state(
+            database,
+            root_hash,
+            tree_depth,
+            dense_prefix_depth,
+            gc_threshold,
+            &initial_leaf_value,
+            &mmap_file_path,
+        )
+        .await?
+        {
+            Some(state) => state,
+            None => match Self::initialize_tree(
+                database,
+                tree_depth,
+                dense_prefix_depth,
+                gc_threshold,
+                initial_leaf_value,
+                mmap_file_path,
+            )
+            .await
+            {
+                anyhow::Result::Ok(state) => state,
+                Err(e) => return Err(e),
+            },
+        };
+
+        Ok(tree_state)
+    }
+
+    #[allow(unused)]
+    async fn get_cached_tree_state(
+        database: &Database,
+        latest_root: ruint::Uint<256, 4>,
+        tree_depth: usize,
+        dense_prefix_depth: usize,
+        gc_threshold: usize,
+        initial_leaf_value: &Hash,
+        mmap_file_path: &String,
+    ) -> anyhow::Result<Option<TreeState>> {
+        let mut mined_items = database.get_commitments_by_status(Status::Mined).await?;
+        let last_mined = if mined_items.is_empty() {
+            initial_leaf_value.clone()
+        } else {
+            mined_items.sort_by_key(|item| item.leaf_index);
+            mined_items.last().map(|item| item.element).unwrap()
```

Comment:
The entire `else` can be replaced with one DB query `get_latest_element`. `There's no need to fetch all mined identities to get one element.

Reply:
I'm not sure if you referred to existing function since i couldn't locate it, so i wrote one. Could you review again now?